File: jbake-core/src/test/java/org/jbake/app/CrawlerTest.java
package org.jbake.app;

import java.util.HashMap;
import java.util.Map;

import org.apache.commons.io.FilenameUtils;
import org.jbake.model.DocumentModel;
import org.jbake.model.DocumentTypes;
import org.jbake.model.ModelAttributes;
import org.jbake.util.DataFileUtil;
import org.junit.jupiter.api.Test;

import com.orientechnologies.orient.core.db.record.OTrackedMap;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;

class CrawlerTest extends ContentStoreIntegrationTest {

    @Test
    void crawl() {
        Crawler crawler = new Crawler(db, config);
        crawler.crawl();

        assertEquals(4, db.getDocumentCount("post"));
        assertEquals(3, db.getDocumentCount("page"));

        DocumentList<DocumentModel> results = db.getPublishedPosts();

        assertThat(results.size()).isEqualTo(3);

        for (Map<String, Object> content : results) {
            assertThat(content)
                    .containsKey(ModelAttributes.ROOTPATH)
                    .containsValue("../../../");
        }

        DocumentList<DocumentModel> allPosts = db.getAllContent("post");

        assertThat(allPosts.size()).isEqualTo(4);

        for (DocumentModel content : allPosts) {
            if (content.getTitle().equals("Draft Post")) {
                assertThat(content).containsKey(ModelAttributes.DATE);
            }
        }

        // covers bug #213
        DocumentList<DocumentModel> publishedPostsByTag = db.getPublishedPostsByTag("blog");
        assertEquals(3, publishedPostsByTag.size());
    }

    @Test
    void crawlDataFiles() {
        Crawler crawler = new Crawler(db, config);
        // manually register data doctype
        DocumentTypes.addDocumentType(config.getDataFileDocType());
        db.updateSchema();
        crawler.crawlDataFiles();
        assertEquals(2, db.getDocumentCount("data"));

        DataFileUtil dataFileUtil = new DataFileUtil(db, "data");
        Map<String, Object> videos = dataFileUtil.get("videos.yaml");
        assertFalse(videos.isEmpty());
        assertNotNull(videos.get("data"));

        // regression test for issue 747
        Map<String, Object> authorsFileContents = dataFileUtil.get("authors.yaml");
        assertFalse(authorsFileContents.isEmpty());
        Object authorsList = authorsFileContents.get("authors");
        assertThat(authorsList).isNotInstanceOf(OTrackedMap.class);
        assertThat(authorsList).isInstanceOf(HashMap.class);
        HashMap<String, Map<String, Object>> authors = (HashMap<String, Map<String, Object>>) authorsList;
        assertThat(authors.get("Joe Bloggs").get("last_name")).isEqualTo("Bloggs");
    }

    @Test
    void renderWithPrettyUrls() {

        config.setUriWithoutExtension(true);
        config.setPrefixForUriWithoutExtension("/blog");

        Crawler crawler = new Crawler(db, config);
        crawler.crawl();

        assertEquals(4, db.getDocumentCount("post"));
        assertEquals(3, db.getDocumentCount("page"));

        DocumentList<DocumentModel> documents = db.getPublishedPosts();

        for (DocumentModel model : documents) {
            String noExtensionUri = "blog/\\d{4}/" + FilenameUtils.getBaseName(model.getFile()) + "/";

            assertThat(model.getNoExtensionUri()).matches(noExtensionUri);
            assertThat(model.getUri()).matches(noExtensionUri + "index\\.html");
            assertThat(model.getRootPath()).isEqualTo("../../../");
        }
    }
}

File: jbake-core/src/test/java/org/jbake/app/CrawlerTest.kt
package org.jbake.app

import com.orientechnologies.orient.core.db.record.OTrackedMap
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.booleans.shouldBeFalse
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import io.kotest.matchers.types.shouldBeInstanceOf
import io.kotest.matchers.types.shouldNotBeInstanceOf
import org.apache.commons.io.FilenameUtils
import org.jbake.model.DocumentModel
import org.jbake.model.DocumentTypes.addDocumentType
import org.jbake.model.ModelAttributes
import org.jbake.util.DataFileUtil

class CrawlerTest : StringSpec({

    lateinit var db: ContentStore
    lateinit var config: org.jbake.app.configuration.DefaultJBakeConfiguration

    beforeSpec {
        ContentStoreIntegrationTest.setUpClass()
        db = ContentStoreIntegrationTest.db
        config = ContentStoreIntegrationTest.config
    }

    beforeTest {
        db.startup()
    }

    afterTest {
        db.drop()
    }

    afterSpec {
        ContentStoreIntegrationTest.cleanUpClass()
    }

    "crawlContentDirectory" {
        val crawler = Crawler(db, config)
        crawler.crawlContentDirectory()

        db.getDocumentCount("post") shouldBe 4
        db.getDocumentCount("page") shouldBe 3

        val results: DocumentList<DocumentModel> = db.publishedPosts

        results.size shouldBe 3

        for (content in results) {
            content.containsKey(ModelAttributes.ROOTPATH) shouldBe true
            content.containsValue("../../../") shouldBe true
        }

        val allPosts: DocumentList<DocumentModel> = db.getAllContent("post")

        allPosts.size shouldBe 4

        allPosts.filter { it.title == "Draft Post" }
            .forEach { it.containsKey(ModelAttributes.DATE) shouldBe true }

        // Covers bug #213
        val publishedPostsByTag: DocumentList<DocumentModel> =
            db.getPublishedPostsByTag("blog")
        publishedPostsByTag.size.toLong() shouldBe 3
    }

    "crawlDataFiles" {
        val crawler = Crawler(db, config)

        // Manually register data doctype.
        addDocumentType(config.dataFileDocType)
        db.updateSchema()

        crawler.crawlDataFiles()
        db.getDocumentCount("data") shouldBe 2

        val dataFileUtil = DataFileUtil(db, "data")
        val videosYaml = dataFileUtil.get("videos.yaml")
        videosYaml.isEmpty().shouldBeFalse()
        videosYaml["data"].shouldNotBeNull()

        // Regression test for issue #747.
        val authorsFileContents = dataFileUtil.get("authors.yaml")
        authorsFileContents.isEmpty().shouldBeFalse()
        val authorsList = authorsFileContents["authors"]
        authorsList.shouldNotBeInstanceOf<OTrackedMap<*>>()
        authorsList.shouldBeInstanceOf<HashMap<*, *>>()
        @Suppress("UNCHECKED_CAST")
        val authors = authorsList as HashMap<String, MutableMap<String, Any>>
        authors["Joe Bloggs"]!!["last_name"] shouldBe "Bloggs"
    }

    "renderWithPrettyUrls" {
        config.setUriWithoutExtension(true)
        config.setPrefixForUriWithoutExtension("/blog")

        Crawler(db, config).crawlContentDirectory()

        db.getDocumentCount("post") shouldBe 4
        db.getDocumentCount("page") shouldBe 3

        val documents: DocumentList<DocumentModel> = db.publishedPosts

        for (model in documents) {
            val noExtensionUri = "blog/\\d{4}/" + FilenameUtils.getBaseName(model.file) + "/"

            model.noExtensionUri!!.matches(noExtensionUri.toRegex()) shouldBe true
            model.uri!!.matches((noExtensionUri + "index\\.html").toRegex()) shouldBe true
            model.rootPath shouldBe "../../../"
        }
    }
})

